src/appleseed/renderer/kernel/lighting/bdpt/bdptlightingengine.cpp

```diff
@@ -90,52 +136,309 @@ namespace
             const ShadingPoint&     shading_point,
             ShadingComponents&      radiance) override      // output radiance, in W.sr^-1.m^-2
         {
-            PathVisitor light_path_visitor;
-            VolumeVisitor volume_visitor;
+            BDPTVertex camera_vertices[8];
+            BDPTVertex light_vertices[9];
 
-            PathTracer<PathVisitor, VolumeVisitor, true> light_path_tracer(     // true = adjoint
-                light_path_visitor,
-                volume_visitor,
-                ~0,
-                1,
-                ~0,
-                ~0,
-                ~0,
-                ~0,
-                shading_context.get_max_iterations());
+            size_t num_light_vertices = trace_light(sampling_context, shading_context, light_vertices);
+            size_t num_camera_vertices = trace_camera(sampling_context, shading_context, shading_point, camera_vertices);
+
+            for (size_t s = 0; s < num_light_vertices + 1; s++)
+                for (size_t t = 2; t < num_camera_vertices + 2; t++)
+                    if (s + t <= num_max_vertices)
+                        connect(shading_context, shading_point, light_vertices, camera_vertices, s, t, radiance);
         }
 
-        void trace_light(
-            SamplingContext&        sampling_context,
-            const ShadingContext&   shading_context)
+        Spectrum compute_geometry_term(
+            const ShadingContext&   shading_context,
+            const ShadingPoint&     shading_point,
+            const BDPTVertex&       a,
+            const BDPTVertex&       b)
+        {
+            double g = 1.0;
+
+            const Vector3d v = b.m_position - a.m_position;
+
+            /// TODO:: the special care have to be taken for these dot products when it comes to volume
+            // compute unnormalied dot products
+            g *= max(-dot(v, b.m_geometric_normal), 0.0);
+            g *= max(dot(v, a.m_geometric_normal), 0.0);
+
+            Spectrum result(0.0);
+
+            // do the visibility test
+            if (g > 0.0)
+            {
+                // compute inverse dist2 and normalize 2 dot products
+                const double dist2 = square_norm(v);
+                g /= (dist2 * dist2);
```

Comment:
You're effectively dividing `g` by `dist^4` here, is that intended?

Reply:
Oh, ok. I think a comment explaining this will go a long way! And/or renaming `g` to `g2`?

Reply:
g (geometry term) was correct. I think the source code might be confusing. It was the trade-off between readability and performance. I'll rewrite the code so that it's more understandable in the next commit.